### PR TITLE
[flaky-ci] Retry MAUI restore during feed publication lag

### PR DESCRIPTION
**Pull request by simonrozsival**
----

## Why

The MAUI integration job can resolve a newly published floating
`Microsoft.Extensions.Logging.Debug` build while NuGet still has an older
cached version list for its exact `Microsoft.Extensions.Logging` dependency.
This produced `NU1102` in builds 1575484, 1581740, 1581742, and 1582480 even
though the requested packages had already reached the `dotnet11` feed.

This addresses https://github.com/dotnet/android/issues/12704.

## What changed

- Pre-restore the generated Android-only MAUI project with `--no-http-cache`
  before the Debug, Release, Mono, and R2R paths.
- Retry only correlated prerelease `Microsoft.Extensions.Logging` `NU1102`
  diagnostics with an exact simple lower-bound requirement where `dotnet11`
  reports an older nearest version.
- Clear only NuGet's HTTP cache between bounded attempts and after an immediate
  success, so later implicit restore graphs cannot reuse stale metadata.
- Keep unrelated, bounded-range, and persistently missing package failures
  gating.
- Use the same helper in the public, internal, and shared MAUI pipelines.

## Validation

- [x] Direct positive and bounded-range negative checks of the restore failure
      classifier.
- [x] Generated Android-only MAUI project restore using the affected
      `Microsoft.Extensions.Logging.Debug` build and the `dotnet11` feed.
- [x] Azure Pipeline YAML syntax parsing.

- [x] Useful description of *why the change is necessary*.
- [x] Links to issues fixed
- [x] Unit tests (not applicable to this CI-only script change)

### CODE REVIEW

File: build-tools/automation/scripts/RestoreMauiTemplate.ps1 [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+		$attemptArguments = @($restoreArguments)
+		if ($BinaryLogPath) {
+			$binaryLogDirectory = Split-Path -Parent $BinaryLogPath
+			if ($binaryLogDirectory -and -not (Test-Path -LiteralPath $binaryLogDirectory)) {
+				New-Item -ItemType Directory -Path $binaryLogDirectory -Force | Out-Null
+			}
+
+			$extension = [IO.Path]::GetExtension($BinaryLogPath)
+			$fileName = [IO.Path]::GetFileNameWithoutExtension($BinaryLogPath)
+			$attemptBinaryLog = Join-Path $binaryLogDirectory "$fileName-attempt-$attempt$extension"
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply build-tools/automation/scripts/RestoreMauiTemplate.ps1:143]: Fixed in 4bd949e6e3. A leaf-only `BinaryLogPath` now uses the current directory before constructing the per-attempt filename, and the PowerShell regression test verifies the resulting `-bl:` argument.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: build-tools/automation/scripts/RestoreMauiTemplate.ps1 [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+			}
+			if ($line -notmatch '(?i)error NU1102: Unable to find package Microsoft\.Extensions\.Logging(?:\.[A-Za-z0-9.-]+)? with version \(>= \d+\.\d+\.\d+-(preview|rc|alpha|beta)[^)]+\)') {
+				return $false
+			}
+
+			$missingPackageCount++
+			$currentPackageHasOlderDotNet11Version = $false
+			continue
+		}
+		if ($missingPackageCount -gt 0 -and $line -match '(?i)Found \d+ version\(s\) in dotnet11 \[ Nearest version: \d+\.\d+\.\d+[^]]* \]') {
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply build-tools/automation/scripts/RestoreMauiTemplate.ps1:51]: Addressed in 2c0bdfcf43 by restricting the accepted NU1102 diagnostic to the exact simple `>= <prerelease>` form. Diagnostics with additional upper bounds or other range constraints remain hard failures, so a newer out-of-range nearest version is no longer classified as publication skew. I also removed the standalone self-test script and its per-build pipeline step; the positive and bounded-range negative cases were validated directly.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: build-tools/automation/scripts/RestoreMauiTemplate.ps1 [hunk 3]
```diff
@@ -0,0 +1,191 @@
+[CmdletBinding()]
+param (
+	[string] $DotNetPath,
+	[string] $Project,
+	[string] $NuGetConfig,
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply build-tools/automation/scripts/RestoreMauiTemplate.ps1:11]: Fixed in 9db99ad6d5. The script now requires non-empty `DotNetPath`, `Project`, and `NuGetConfig` values at both the script entry point and helper boundary. I also removed the unused `DefineOnly` mode. Empty input now fails during parameter binding with the parameter name.

File: build-tools/automation/scripts/RestoreMauiTemplate.ps1 [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+		$output | ForEach-Object { Write-Host "$_" }
+		$lastExitCode = [int] $result.ExitCode
+		if ($lastExitCode -eq 0) {
+			if (-not $httpCacheCleared) {
+				Write-Host "Clearing NuGet's HTTP cache so subsequent MAUI restore graphs cannot reuse stale package metadata."
+				$clearExitCode = Clear-NuGetHttpCache -CommandInvoker $CommandInvoker
+				if ($clearExitCode -ne 0) {
+					Write-Host "NuGet HTTP cache clearing failed with exit code $clearExitCode."
+					return $clearExitCode
+				}
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply build-tools/automation/scripts/RestoreMauiTemplate.ps1:159]: I am keeping this as a hard failure because cache invalidation is part of the readiness gate, not unrelated cleanup. `--no-http-cache` lets this restore bypass stale metadata but does not make the existing HTTP cache safe for the later Debug, Release, Mono, and R2R restore graphs, which do not use that option. If clearing fails and this step passes, those later restores can reuse the stale package index and reproduce the original NU1102. Surfacing the cache-clear exit code preserves the gate guarantee and gives a direct diagnostic.
